text/0010-bft-signatures.md

```diff
@@ -0,0 +1,291 @@
+
+```
+  // SeekContentType indicates what type of content to deliver in response to a request. If BLOCK is specified,
+  // the orderer will stream blocks back to the peer. This is the default behavior. If HEADER_WITH_SIG is  specified, the
+  // orderer will stream only a the header and the signature, and the payload field will be set to nil. This allows
+  // the requester to ascertain that the respective signed block exists in the orderer (or cluster of orderers).
+  enum SeekContentType {
+      BLOCK = 0;
+      HEADER_WITH_SIG =1;
+   }
+```
+
+By making the default value be `BLOCK` the API is backward compatible to old consumers.
+
+Indeed, the SmartBFT's approach for Byzantine fault tolerant block replication overcomes the linear overhead of the naive approach. However, a malicious ordering node can still withhold a block up until right before the suspicion timeout expires at the party that pulls the block.
+Hence, while a malicious orderer may not withhold blocks indefinitely, it may slow down the replication of its consumers.
+
+Another minor inconvenience is that the peer admin needs to configure the peer correctly to use the Byzantine fault tolerant implementation instead of the regular one.
+
+### Signature identity referencing
+
+In SmartBFT, like Raft, each configuration block contains a set of consenter (ordering service consensus protocol participant) definitions. 
+Each such consenter definition contains various information that is used either to connect or to authenticate the consenter:
+
+```
+message Consenter {
+    uint64 consenter_id = 1;
+    string host = 2;
+    uint32 port = 3;
+    string msp_id = 4;
+    bytes identity = 5;
+    bytes client_tls_cert = 6;
+    bytes server_tls_cert = 7;
+}
+```
+
+When validating a signature signed by orderer with numerical identifier **j**, the mapping between **j** and the matching identity (which carries the public key) is taken from the latest configuration block.
+
+Recall, the naive approach to encode several signatures on a block in its metadata involves encoding an amount of orderer identities that is linear in the number of ordering nodes.
+
+To that end, SmartBFT substitutes the identity (over 700B in size) with its consensus node identifier (a 64 bit number). 
+As a result, the `MetadataSignature` now contains three additional fields:
+
+```
+message MetadataSignature {
+    bytes signature_header = 1; // An encoded SignatureHeader
+    bytes signature = 2;       // The signature over the concatenation of the Metadata value bytes, signatureHeader, and block header
+    uint64 signer_id = 3;
+    bytes nonce = 4;
+}
+```
+
+When a Fabric node validates a signature on a block, an ephemeral `SignatureHeader` is constructed where the `nonce` is taken from the
+`MetadataSignature` and the `creator` is the `identity` matching the `signer_id`. 
+
```

Comment:
Not related to this RFC along, in general, I know that we include the identity as part of the value for the signatures but not sure if enhances security by any means? Can we not simply live without it and hence the consenter mapping?

Reply:
The fact that we sign over the public key is irrelevant.

You still need to know which public key to use to validate the signature, right? When sending signatures and storing them on storage you need either to mark who are the signers or to simply attach the identity to the signature.

If you attach the identity to the signature, then you get a high storage overhead. 

If you don't attach the identity, then you need to know what identities are possible candidates. But where are you going to take them from? It's tempting to take the block signature policy and convert it to identities and then try to match a candidate MSP principal to a signature but what if the policy is going to be with role MSP principals? then you won't be able to do that anymore.

Reply:
> The fact that we sign over the public key is irrelevant.

The relevance of this was, quite a good part part of you write up was around how to avoid including the cert and my question was can we get rid of this all together?

yes, of course you need the hint of Ids!